web/pandas/pdeps/0004-consistent-to-datetime-parsing.md

```diff
@@ -0,0 +1,88 @@
+# PDEP-4: Consistent datetime parsing
+
+- Created: 18 September 2022
+- Status: Under discussion
+- Discussion: [#48621](https://github.com/pandas-dev/pandas/pull/48621)
+- Author: [Marco Gorelli](https://github.com/MarcoGorelli)
+- Revision: 1
+
+## Abstract
+
+The suggestion is that:
+- ``to_datetime`` becomes strict and uses the same datetime format to parse all elements in its input.
+  The format will either be inferred from the first non-NaN element (if `format` is not provided by the user), or from
+  `format`;
+- ``infer_datetime_format`` be deprecated (as a strict version of it will become the default);
+- an easy workaround for non-strict parsing be clearly documented.
+
+## Motivation and Scope
+
+Pandas date parsing is very flexibible, but arguably too much so - see
+https://github.com/pandas-dev/pandas/issues/12585 and linked issues for how
+much confusion this causes. Pandas can swap format midway, and though this
+is document, it regularly breaks users' expectations.
+
+Simple example:
+```ipython
+In [1]: pd.to_datetime(['12-01-2000 00:00:00', '13-01-2000 00:00:00'])
+Out[1]: DatetimeIndex(['2000-12-01', '2000-01-13'], dtype='datetime64[ns]', freq=None)
+```
+The user was almost certainly intending the data to be read as "12th of January, 13th of January".
+However, it's read as "1st of December, 13th of January". No warning or error is thrown.
+
+Currently, the only way to ensure consistent parsing is by explicitly passing
+``format=``. The argument ``infer_datetime_format``
+isn't strict and can still break users' expectations.
+
+```ipython
+In [2]: pd.to_datetime(['12-01-2000 00:00:00', '13-01-2000 00:00:00'], infer_datetime_format=True)
+Out[2]: DatetimeIndex(['2000-12-01', '2000-01-13'], dtype='datetime64[ns]', freq=None)
+```
+
+## Detailed Description
+
+Concretely, the suggestion is:
+- if no ``format`` is specified, ``pandas`` will guess the format from the first non-NaN row
+  and parse the rest of the input according to that format. Errors will be handled
+  according to the ``errors`` argument - there will be no silent switching of format;
+- ``infer_datetime_format`` will be deprecated;
```

Comment:
If you were parsing `11-12-10`, would the default be November 12, 2010, or December 11, 2010 or December 10, 2011?  I think we should be explicit in this proposal on how `dayfirst` and `yearfirst` interacts with this particular behavior, especially since the docs say that those 2 parameters are not strict.

Reply:
I dont know what `infer_datetime_format` currently does but is it not very restrictive, and not always possible, to infer from just the first element?

If that were to be the case I support the comment of outlining how to operate with year first and day first. As a European my colleagues and I hate US mm/dd/yy and I think this might also need outlining some specifics.

The advantage of providing a consistent format input is that better inference could be made from multiple samples and this is very useful for structured data.

Reply:
Thanks both for taking a look!

@Dr-Irv this proposal wouldn't change how `dayfirst` and `yearfirst` operate. The format will try to be guessed in accordance with these parameters, just like it is on `main` - the difference is that with this proposal, the format guessed from the first non-NaN row will be used to parse the rest of the Series

@attack68 in the rare case that it's not possible to guess the format from the first element, then a UserWarning would be raised, check lines 49-55 of this PR

You're very right to bring up `mm/dd/yy` 👍  - indeed the vast majority of the world doesn't use that format. That's why the current behaviour is so dangerous. For example, suppose your data is in `%d-%m-%Y %H:%M` format:

On `main`, the first row's date would be parsed as `mm-dd-yyyy`, whilst the second one as `dd-mm-yyyy`. No error, no warning, this is very easy to miss (and I almost did once in a prod setting 😳 ):
```ipython
In [1]: pd.to_datetime(['12-01-2000 00:00', '13-01-2000 00:00'])
Out[1]: DatetimeIndex(['2000-12-01', '2000-01-13'], dtype='datetime64[ns]', freq=None)
```

With this PDEP, you could just check the format of your first row, and you'd know the rest of the Series was parsed in accordance to that. If it can't be, then with `errors='raise'` (the default), you'd get an error
```
ValueError: time data '13-01-2000 00:00' does not match format '%m-%d-%Y %H:%M' (match)
```
and you'd see that the guessed format wasn't right. You could get around that either by explicitly passing `format`, or with `dayfirst=True`:
```ipython
In [2]: pd.to_datetime(['12-01-2000 00:00', '13-01-2000 00:00'], dayfirst=True)
Out[2]: DatetimeIndex(['2000-01-12', '2000-01-13'], dtype='datetime64[ns]', freq=None)
```

Totally agree on better documenting this, and that inference could be optimised by using multiple samples to guess - first, I just wanted to get agreement that we want `to_datetime` to parse consistently